opm/autodiff/CPRPreconditioner.hpp

```diff
@@ -43,8 +46,222 @@
 
 namespace Opm
 {
+namespace
+{
+//! \brief A custom deleter for the parallel preconditioners.
+//!
+//! In dune-istl they hold a reference to the sequential preconditioner.
+//! In CPRPreconditioner we use unique_ptr for the memory management.
+//! Ergo we need to construct the sequential preconditioner with new and
+//! make sure that it gets deleted together with the enclosing parallel
+//! preconditioner. Therefore this deleter stores a pointer to it and deletes
+//! it during destruction.
+template<class PREC>
+class ParallelPreconditionerDeleter
+{
+public:
+    ParallelPreconditionerDeleter()
+        : ilu_()
+    {}
+    ParallelPreconditionerDeleter(PREC& ilu)
+    : ilu_(&ilu){}
+    template<class T>
+    void operator()(T* pt)
+    {
+        if(ilu_)
```

Comment:
This seems to be a recurring pattern, I wonder why. I thought perhaps free() was not null safe, but it is, too.